Fix incorrect `transactional()` handling when DB auto-rolled back the transaction

simPod opened the pull request:
<!-- Fill in the relevant information below to help triage your pull request. -->

|      Q       |   A
|------------- | -----------
| Type         | bug
| Fixed issues | #3423

#### Summary

_https://github.com/doctrine/dbal/pull/6543 partially solved what we tried to achieve in https://github.com/doctrine/dbal/pull/4846. With that `There's no active transaction` exception contains the actual exception in `$previous`._

Now, let's get rid of `There's no active transaction` exception which occurs e.g. when using deferred constraints so the violation is checked at the end of the transaction and not during it. Database then rolls back the transaction automatically.

The current faulty flow is following: 

1. on `commit()`, Constrain violation occurs
2. transaction is implicitly aborted and rolled back in database
3. DBAL calls `rollback()`
4. Since there's no active transaction, `There's no active transaction` exception is thrown:

<img width="1202" alt="image" src="https://github.com/user-attachments/assets/c5c7d77f-c783-4902-ad28-689107c6a338">

This PR adjusts the flow so:

1. on `commit()`, Constrain violation occurs
2. transaction is implicitly aborted and rolled back in database
3. DBAL **DOES NOT** call `rollback()` 🥳 

-----------

For DBAL v4 it looks like this https://github.com/doctrine/dbal/pull/6546 (draft, does not contain review changes)

simPod committed:
Fix incorrect `transactional()` handling when DB auto-rolled back the transaction

Let's get rid of There's no active transaction exception which occurs e.g. when using deferred constraints so the violation is checked at the end of the transaction and not during it.

- Do not rollback only on exceptions where we know the transaction is no longer active
- Introduce TransactionRolledBack exception
- Transform Oracle's "transaction rolled back" exception and use the underlying one that DBAL supports

greg0ire commented on src/Connection.php:
```diff
@@ -1283,16 +1283,17 @@ public function transactional(Closure $func)
             }
         }
+
+        $this->commit();
```
What if `commit` throws for some other reason, before anything is transmitted to the database?

simPod replied:
Can you describe some scenario you have in mind so I can cover it with test?

greg0ire replied:
Here, there are throws statements:

https://github.com/doctrine/dbal/blob/220aec46be990c0900826e29d9e3faa53bfbf70a/src/Connection.php#L1013-L1026

I don't know if there is a realistic scenario in which they can throw when `commit` is called from the method you are modifying, before the transaction is actually committed, but that's what I was thinking about.

simPod replied:
I see. Those should not happen but I have adjusted the flow and covered with unit test should it happen.

72af9b864068e140d0f67186cfa7dc201a3244ac

greg0ire commented on src/Connection.php:
```diff
@@ -9,6 +9,7 @@
 use Doctrine\DBAL\Cache\QueryCacheProfile;
 use Doctrine\DBAL\Driver\API\ExceptionConverter;
 use Doctrine\DBAL\Driver\Connection as DriverConnection;
+use Doctrine\DBAL\Driver\Exception as TheDriverException;
```
Why this `The` prefix? We didn't use it when aliasing `DriverConnection`.

simPod replied:
There's already `use Doctrine\DBAL\Exception\DriverException;` imported on L19

simPod replied:
And now it's gone https://github.com/doctrine/dbal/pull/6545#discussion_r1797719214

greg0ire commented on src/Connection.php:
```diff
@@ -1283,16 +1284,30 @@ public function transactional(Closure $func)
+            if ($shouldRollback) {
+                $this->rollBack();
+            }
+        }
```
Why not do something similar to what I did in https://github.com/doctrine/orm/pull/11646/commits/b6137c89110952664cdc0b2e4df7f98716ff3f93 and use `isTransactionActive()`?

simPod replied:
I forgot about that API, should work.

greg0ire replied:
What should happen in this scenario?

```php
$connection->beginTransaction();
$connection->transactional(some function call here);
$connection->commit();
```

I'm not familiar with nested transactions, but if it's possible to nest them then this will return true because of the outer transaction, although the outer transaction will have been property closed. Also, why not reuse the previous `try` / `finally` block just like I did in ORM?

simPod replied:
Nested transactions should be covered by 4 tests.

I merged try blocks, it makes sense and is nice with the `isTransactionActive` API.

greg0ire replied:
Now the code looks like it is going to wrongly revert the outer transaction after successfully committing the inner one, just because the outer one is active (as it should)

greg0ire replied:
I'm surprised the test suite is green, maybe I'm missing something.

simPod replied:
Haha that's funny. I'll add a test case to verify it but I think you're right.

simPod replied:
Umm I'm confused now. Will take a break.

simPod replied:
I can add a check of nesting level but the green test suite somehow implies it won't happen at all anyway.

simPod replied:
It indeed tried to rollback the transaction. I have reverted to a previous version and added a test to cover the case.

greg0ire replied:
I am confused now. Are you saying that the bug exists, and that you reproduced it? I don't see any check on the transaction nesting level, which does sound exactly like what would be needed here, so why is the tests suite mostly green (it's red on Oracle, do you know why?)

simPod replied:
> it's red on Oracle, do you know why

oracle and db2 require `select...` to have also `from` clause. Fixed.

> Are you saying that the bug exists, and that you reproduced it

Yes, the bug would have been introduced with 88879b3ad5cabff479e1e85a2975091887f2744e. I covered it with `testNestedTransactionWalkthrough()` so we know it passes.

When doing `commit()`, we need to rollback the transaction only if something failed outside the database https://github.com/doctrine/dbal/pull/6545#discussion_r1797695417.

Since we need to have `rollback()` in `finally`, we need to use boolean flag to know whether to trigger it.  
We can't use `isTransactionActive()` since it only checks the nesting level. `isTransactionActive()` can actually be renamed to something `isTransactionNested()`.

greg0ire commented:
It seems there are CI jobs failing. Please take a look at [this guide](https://www.doctrine-project.org/contribute/#dealing-with-checks-and-tools) for more on how to handle those.

simPod commented:
Thanks, just did not notice that basic stuff, u were faster 🙃 fixed.

greg0ire commented on src/Connection.php:
```diff
@@ -1424,12 +1439,21 @@ public function commit()
+                $this->releaseSavepoint($this->_getNestedTransactionSavePointName());
+            }
+        } finally {
+            $this->updateTransactionStateAfterCommit();
```
If the above lines are successful but something goes wrong here, then `$shouldRollback` will be `true` (right?) although no rollback should happen.

That's why I was expecting some code that takes a look at the transaction nesting level before and after the commit. If the transaction level decreased all by itself.

Looking at the code though, I don't see how it could fail, so I guess this is OK.

greg0ire replied:
It's unclear to me why you are calling this function inside a `finally` block. If the commit fails, why should the transaction nesting level decrease?

simPod replied:
> If the commit fails, why should the transaction nesting level decrease?

We need to cleanup the transaction state. Otherwise `isTransactionActive()` will still report true to other functions (e.g. in `disconnect()`) which is not true actually. It is covered by UniqueConstraintViolationsTest.

greg0ire replied:
Why should it no longer be considered active? If `COMMIT` fails, then the transaction will be "aborted", but it won't be auto-rolled-back, right? So to me, I'd say the transaction is still going on, but it is aborted and unusable, and to regain control, you have to use `ROLLBACK TO` or `ROLLBACK`.

That's what I understand after reading the last few sentences of  https://www.postgresql.org/docs/current/tutorial-transactions.html anyway.

simPod replied:
You're right, good point.

I came up with the solution where we can check the exception type and if it is a known case where the transaction is auto rolled back, we won't call ROLLBACK.

[36e23d1](https://github.com/doctrine/dbal/pull/6545/commits/36e23d1f7d76081f7bb3337b7399172a419dcafe)

simPod commented:
~_I'll have to find a way to install php OCI drivers on m1 macbook to debug oracle failures_~  
Solved

greg0ire commented:
I still don't get why you don't simply do this:

```php
$successful = false;
try {
    $res = $func($this);
    $this->commit();

    $successful = true;

    return $res;
} finally {
    if (! $successful && $this->isTransactionActive()) {
        $this->rollBack();
    }
}
```

simPod commented:
> I still don't get why you don't simply do this:
> 
> ```
> $successful = false;
> try {
>     $res = $func($this);
>     $this->commit();
> 
>     $successful = true;
> 
>     return $res;
> } finally {
>     if (! $successful && $this->isTransactionActive()) {
>         $this->rollBack();
>     }
> }
> ```

isTransactionActive() has actually no idea whether the transaction is, active. We need to check the actual exception from db so we know whether it was automatically rolled back or not.

greg0ire commented:
>  it was automatically rolled back

Is that a thing that actually can happen? Also, I think if the commit succeeds, but the subsequent https://github.com/doctrine/dbal/blob/220aec46be990c0900826e29d9e3faa53bfbf70a/src/Connection.php#L1038 fails, then you don't want to rollback, and checking `isTransactionActive` should help with that scenario.

simPod commented:
> Is that a thing that actually can happen?

Yes, that's why we're getting `There's no active transaction` exceptions. It's covered in the added tests. Those exceptions come from the database itself when it evaluates deferred constraints during commit. If such commit fails and you call explicit `rollback`, you get the `There's no active transaction` error and that's what I'm trying to solve here.

Oracle has specific error for that case: 2091. It is implemented in f215f4b7a3bdf4db56f7cfa017609266b3babaa9

> Also, I think if the commit succeeds, but the subsequent `beginTransaction`() fails, then you don't want to rollback

I'm not sure I understand what you mean. Can you maybe provide an sql example of the scenario? I'd like to cover every doubt with a test.

`isTransactionActive()` is IMO not helpful with deferred constraints.

greg0ire commented:
> Can you maybe provide an sql example of the scenario?

I will try

```sql
BEGIN TRANSACTION;
-- our SQL goes here
COMMIT; -- success!
BEGIN TRANSACTION; -- this one happens when auto-commit is set to false and transaction nesting level is back to 0, and may fail for various reasons, I suppose
```

simPod commented:
> -- this one happens when auto-commit is set to false and transaction nesting level is back to 0, and may fail for various reasons, I suppose

I just checked and don't see how it can happen (but maybe I'm missing something?).

I have added 2 tests to cover how `transactional()` behaves in different auto commit modes.

`testTransactionIsNotActiveAfterTransactionalInAutoCommitMode()`  
`testTransactionIsActiveAfterTransactionalInNoAutoCommitMode()`
  
When `autocommit=false`, then calling `transactional()` increases nesting level to `2` and this line is never called https://github.com/doctrine/dbal/blob/220aec46be990c0900826e29d9e3faa53bfbf70a/src/Connection.php#L1038.

This PR seems to have no effect on scenarios outside of `transactional()` https://github.com/doctrine/dbal/pull/6545#issuecomment-2412377607, rollback is not called.

greg0ire commented:
> I just checked and don't see how it can happen (but maybe I'm missing something?).

Network error? Database issue caused by some other program? I don't see clearly how it could happen either. If there are tests that can be extracted to a separate PR, please do. Otherwise I fear that when I'm finally satisfied with all this and ask for other maintainers to review, there will be more discussions.

Another reason is because [it might not be very straightforward to merge up](https://github.com/doctrine/dbal/pull/6555)

simPod commented:
> how it can happen

> Network error? Database issue caused by some other program?

yes, of course. I meant "how the wrong rollback" can happen in this code flow.

>  If there are tests that can be extracted to a separate PR, please do

sure, I'll try to minimize the PR. I expect long discussions anyway 🙃 .

greg0ire commented:
> yes, of course. I meant "how the wrong rollback" can happen in this code flow.

Let's say the network error happens and the opening of the transaction fails. An exception will bubble up, preventing `$successful` from switching to `true`, and there will be an attempt to rollback a transaction that was never happened in the first place. Right?

simPod commented:
> > yes, of course. I meant "how the wrong rollback" can happen in this code flow.
> 
> Let's say the network error happens and the opening of the transaction fails. An exception will bubble up, preventing `$successful` from switching to `true`, and there will be an attempt to rollback a transaction that was never happened in the first place. Right?

I think I covered the case where 

`transactional()` is used here

https://github.com/doctrine/dbal/pull/6545/files#diff-a3bde44a40c3047e1a8c316dd4e5dbb1c8f8d169fa9558e01faabbb2a794d58cR400-R440

and where `transactional()` is not used here 

https://github.com/doctrine/dbal/pull/6545/files#diff-a3bde44a40c3047e1a8c316dd4e5dbb1c8f8d169fa9558e01faabbb2a794d58cR459-R486

---------

> When autocommit=false, then calling transactional() increases nesting level to 2 and this line is never called

therefore, there does not appear to be any attempt to rollback 🤔

greg0ire commented:
> > When autocommit=false, then calling transactional() increases nesting level to 2 and this line is never called
> therefore, there does not appear to be any attempt to rollback 🤔

Makes sense to me. What about when `autocommit=true`? Sorry if this should be obvious to me, but I don't have all the details loaded in mind.

simPod commented:
> > > When autocommit=false, then calling transactional() increases nesting level to 2 and this line is never called
> > > therefore, there does not appear to be any attempt to rollback 🤔
> 
> Makes sense to me. What about when `autocommit=true`? Sorry if this should be obvious to me, but I don't have all the details loaded in mind.

Absolutely understand.

I have covered calling `transactional()` with and without autocommit in those tests.

> What about when autocommit=true

When it's true it does not satisfy the left side of the condition in order to rollback

https://github.com/simPod/dbal/blob/e6cf572874bf74f816b94f034be9440aed1c8d76/src/Connection.php#L1482

If it's false it does not satisfy the right side because

> When autocommit=false, then calling transactional() increases nesting level to 2 and this line is never called.

simPod commented:
I think I have addressed @/derrabus' (not tagging explicitly to avoid noise) comments from previous attempt related to OCI and PDO_OCI https://github.com/doctrine/dbal/pull/4846#pullrequestreview-1761176225

greg0ire commented on tests/ConnectionTest.php:
```diff
@@ -396,6 +397,48 @@ public function testConnectStartsTransactionInNoAutoCommitMode(): void
+    public function testTransactionIsNotActiveAfterTransactionalInAutoCommitMode(): void
+    {
+        $driverMock = $this->createMock(Driver::class);
+        $driverMock->expects(self::any())
```
I don't think `->expects(self::any())` achieves anything, it can be removed IMO

```suggestion
        $driverMock
```

There are other occurrences below.

simPod replied:
> ~Required parameter '$invocationRule' missing~

greg0ire commented on tests/ConnectionTest.php:
```diff
@@ -396,6 +397,48 @@ public function testConnectStartsTransactionInNoAutoCommitMode(): void
 
+    public function testTransactionIsNotActiveAfterTransactionalInAutoCommitMode(): void
+    {
+        $driverMock = $this->createMock(Driver::class);
```
That's a stub since you are not asserting anything on it.

```suggestion
        $driverMock = $this->createStub(Driver::class);
```

simPod replied:
I'm not familiar with phpunit, I copied it from other tests. Changed to stub, thanks.

greg0ire commented on tests/Functional/ForeignKeyConstraintViolationsTest.php:
```diff
@@ -0,0 +1,288 @@
+//            $this->expectException(DriverException::class);
+//
+//            return;
+//        }
```
Should this PR be in draft status then?

simPod replied:
Missed this, resolved.

greg0ire commented:
Please use an interactive rebase to make your commits atomic (all commits should be "revertable" / pass the build). Also, if you can split this PR, that would be great.

simPod commented:
Yup, though for the review process I wanted the commits to represent addressing the comments / explain the issue.

Will consolidate + continue extracting.

greg0ire commented on src/Exception/TransactionRolledBack.php:
```diff
@@ -0,0 +1,24 @@
+        $this->why = $why;
+    }
+
+    public function why(): DriverException
```
Is this method called anywhere?

simPod replied:
Nope but I'd keep it as a public api to be able to fetch the cause of the exception

greg0ire replied:
I don't think users should have anything to do to see the exception. Maybe `$why` and `$exception` should be chained together, just like what happens when having exception in `try` and the corresponding `finally`?

simPod replied:
Good point, now I see it can be simplified.

simPod replied:
(the CI error does not seem relevant)

morozov commented:
It will take me some time to understand the details in order to provide a meaningful feedback. Please feel free to merge it as is. The number of "line not covered by tests" warnings and enumeration of specific exception classes in handling a commit failure is a bit concerning, though. On the positive side, the fix doesn't introduce any new API, so if something goes seriously wrong, it will be safe to revert.

simPod commented:
I guess there's some issue with coverage configuration since e.g. my test explicitly covers this line

<img width="875" alt="image" src="https://github.com/user-attachments/assets/359a823f-ee93-48c9-9d44-2655677a667a">

<img width="1151" alt="image" src="https://github.com/user-attachments/assets/cc807e17-8ae1-4105-bcef-a360ec7fbeff">

greg0ire commented:
I think that for some reason, codecov only processes the reports from appveyor, and not from github-action. I have yet to figure out why.

greg0ire commented:
Thanks @simPod :tada:

simPod commented:
Let me know if you need help resolving v4 conflicts.

greg0ire commented:
Thanks for offering your help, I do need it :pray:
